### PR TITLE
test/osd: test environment for the scrubber "backend"

### PR DESCRIPTION
**Pull request by ronen-fr**
A framework for unit-testing the scrubber code:

Introducing the scrubber_generators to create
scrubber test data, and the scrubber_test_datasets
for pre-prepared test configurations.

First tests of the scrubber backend (in test_scrubber_be.cc).

Including some required changes to the scrubber itself, further
minimizing its external interfaces.

Signed-off-by: Ronen Friedman <rfriedma@redhat.com>

### CODE REVIEW

File: src/osd/scrubber/scrub_backend.cc [hunk 1]
```diff
@@ -40,24 +40,22 @@ std::ostream& ScrubBackend::logger_prefix(std::ostream* out,
 // ////////////////////////////////////////////////////////////////////////// //
 
 // for a Primary
-ScrubBackend::ScrubBackend(PgScrubber& scrubber,
-                           PGBackend& backend,
-                           PG& pg,
+ScrubBackend::ScrubBackend(ScrubBeListener& scrubber,
+                           PgScrubBeListener& pg,
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.cc:44]: Why are these separate?

File: src/osd/scrubber/scrub_backend.cc [hunk 2]
```diff
@@ -311,6 +307,13 @@ void ScrubBackend::update_authoritative()
   }
 }
 
+#ifdef NOTYET
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.cc:310]: For the actual PR, please remove this bit from here and simply add a comment to where it ends up that you'd like to move it back.

File: src/osd/scrubber/scrub_backend.cc [hunk 3]
```diff
@@ -867,12 +872,11 @@ std::optional<std::string> ScrubBackend::compare_obj_in_maps(
...
-    for_empty_auth_list(std::move(auths),
-                        std::move(objerrs),
-                        auth,
-                        ho,
-                        errstream);
+  auto opt_ers = for_empty_auth_list(std::move(auths),
+                                     std::move(objerrs),
+                                     auth,
+                                     ho,
+                                     errstream);
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.cc:879]: Unnecessary change.

File: src/osd/scrubber/scrub_backend.cc [hunk 4]
```diff
@@ -1282,7 +1286,9 @@ bool ScrubBackend::compare_obj_details(pg_shard_t auth_shard,
     auth_bl.push_back(auth_attr->second);
 
     if (!can_bl.contents_equal(auth_bl)) {
-      format_to(std::back_inserter(out), "{}object info inconsistent ", sep(error));
+      format_to(std::back_inserter(out),
+                "{}object info inconsistent ",
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.cc:1290]: Space between {} and object

Comment (athanatos) [reply src/osd/scrubber/scrub_backend.cc:1290]: Also, unnecessary change?

File: src/osd/scrubber/scrub_backend.cc [hunk 5]
```diff
@@ -1302,7 +1308,9 @@ bool ScrubBackend::compare_obj_details(pg_shard_t auth_shard,
       auth_bl.push_back(auth_attr->second);
 
       if (!can_bl.contents_equal(auth_bl)) {
-        format_to(std::back_inserter(out), "{}snapset inconsistent ", sep(error));
+        format_to(std::back_inserter(out),
+                  "{}snapset inconsistent ",
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.cc:1312]: Unnecessary change?

File: src/osd/scrubber/scrub_backend.h [hunk 6]
```diff
@@ -46,15 +46,17 @@
...
 struct ScrubMap;
 
 class PG;
 class PgScrubber;
-class PGBackend;
-class PGPool;
-
+//class PGBackend;
+struct PGPool;
+using Scrub::PgScrubBeListener;
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.h:59]: I'm not sure why this needs to be separate from ScrubBeListener -- they only seem to be consumed by ScrubBackend.

File: src/osd/scrubber_common.h [hunk 7]
```diff
@@ -12,6 +12,24 @@ namespace ceph {
 class Formatter;
 }
 
+struct PGPool;
+
+namespace Scrub {
+  class ReplicaReservations;
+}
+
+/// Facilitating scrub-realated object access to private PG data
```
Comment (athanatos) [inline src/osd/scrubber_common.h:21]: Is this actually useful?  Seems like you could simply declare *Listener to be a friend of PG without much loss of expressiveness.

File: src/osd/osd_types_fmt.h [hunk 8]
```diff
@@ -16,8 +16,11 @@ struct fmt::formatter<osd_reqid_t> {
   template <typename FormatContext>
   auto format(const osd_reqid_t& req_id, FormatContext& ctx)
   {
-    return fmt::format_to(ctx.out(), "{}.{}:{}", req_id.name, req_id.inc,
-			  req_id.tid);
+    return fmt::format_to(ctx.out(),
+                          "{}.{}:{}",
+                          req_id.name,
+                          req_id.inc,
+                          req_id.tid);
```
Comment (athanatos) [inline src/osd/osd_types_fmt.h:23]: Unnecessary?

File: src/osd/osd_types_fmt.h [hunk 9]
```diff
@@ -56,9 +59,12 @@ struct fmt::formatter<chunk_info_t> {
...
   {
-    return fmt::format_to(ctx.out(), "(len: {} oid: {} offset: {} flags: {})",
-			  ci.length, ci.oid, ci.offset,
-			  ci.get_flag_string(ci.flags));
+    return fmt::format_to(ctx.out(),
+                          "(len: {} oid: {} offset: {} flags: {})",
+                          ci.length,
+                          ci.oid,
+                          ci.offset,
+                          ci.get_flag_string(ci.flags));
```
Comment (athanatos) [inline src/osd/osd_types_fmt.h:67]: Unnecessary?

Comment (athanatos) [reply src/osd/osd_types_fmt.h:67]: There seem to be a bunch of these mixed into behavior related changes.  Either remove them or separate them into their own commit.

File: src/osd/osd_types_fmt.h [hunk 10]
```diff
@@ -132,3 +146,153 @@ struct fmt::formatter<spg_t> {
...
+      for (const auto& [ho, so] : smap.objects) {
+        fmt::format_to(ctx.out(), "\n\th.o<{}>:<{}> ", ho, so);
+      }
+      fmt::format_to(ctx.out(), "\n");
+    }
+    return fmt::format_to(ctx.out(), "}}");
+  }
+
+  bool debug_log{false};
+};
```
Comment (athanatos) [inline src/osd/osd_types_fmt.h:298]: End of file newline.

File: src/test/osd/test_scrubber_be.cc [hunk 11]
```diff
@@ -0,0 +1,870 @@
...
+
+    sbe->new_chunk();
+    std::cout << fmt::format("{}: TestScrubBackend created\n", __func__);
+    fake_a_scrub_set(real_objs_list);
+    std::cout << fmt::format("{}: fake_a_scrub_set() done\n", __func__);
+  }
+
+  void TearDown() override;
+
+  // void fake_a_scrub_set(ScrubGenerator::RealObjsConf conf);
```
Comment (athanatos) [inline src/test/osd/test_scrubber_be.cc:303]: Instead of this comment, indicate what the function actually does.

File: src/test/osd/test_scrubber_be.cc [hunk 12]
```diff
@@ -0,0 +1,870 @@
...
+
+  pg_info_t setup_pg_in_map();
+};
+
+
+// ///////////////////////////////////////////////////////////////////////////
+// ///////////////////////////////////////////////////////////////////////////
+
+
+// copied from TestOSDMap.cc
```
Comment (athanatos) [inline src/test/osd/test_scrubber_be.cc:346]: Probably worth separating out into a common header.

File: src/test/osd/test_scrubber_be.cc [hunk 13]
```diff
@@ -0,0 +1,870 @@
...
+  sbe->update_repair_status(true);
+  ASSERT_TRUE(sbe->get_m_repair());
+
+  // am "I" the first in the acting set?
+  auto others = sbe->all_but_me();
+  auto in_others = std::find(others.begin(), others.end(), i_am);
+  ASSERT_EQ(others.end(), in_others);
+}
+
+
```
Comment (athanatos) [inline src/test/osd/test_scrubber_be.cc:850]: Probably worth adding a one sentence explanation of what each test is meant to do.

**Comment by athanatos** [comment]
Overall strategy seems fine, I'll take another look once it's ready for review.

File: src/osd/scrubber/scrub_backend.cc (hunk 1 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.cc:44]: The services supplied by the PgScrubber vs. direct requests from the PG (that have no equivalent in the PgScrubber).

**Comment by ronen-fr** [comment]
@athanatos - thanks for the review. I will create a cleaner (and more functional) version.
I also plan to move the logger changes into a separate PR.

File: src/osd/scrubber/scrub_backend.cc (hunk 3 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.cc:879]: The formatter forced 80-columns limit on the touched files. Will move all those changes to
a specific PR.
And - BTW - I wish we reconsider and allow 85 or 90 char lines. Too many line breaks this way.

File: src/osd/scrubber/scrub_backend.cc (hunk 4 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.cc:1290]: Adding the space will change existing logs. Might affect tests. Not part of this PR.
Re the formatting - a result of forcing  80-cols. Will move elsewhere. See the general comment below.

File: src/osd/scrubber/scrub_backend.h (hunk 6 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.h:59]: But supplied by two different entities.

File: src/osd/scrubber/scrub_backend.cc (hunk 3 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_backend.cc:823]: There's a great deal of intrinsic value in the whole project using the same formatting norms.  I don't think that changing to 85 or 90 character lines is at all worth breaking that property.

Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.cc:823]: I agree. I meant allowing that for the whole project. I'd suggest that modern monitors are wide enough to hold multiple 90-cols windows.

Comment (athanatos) [reply src/osd/scrubber/scrub_backend.cc:823]: That's a discussion for ceph-devel then, but fwiw, I really prefer 80 chars because it allows me to fit more files on a single monitor horizontally.

File: src/osd/scrubber_common.h (hunk 7 above)
Comment (ronen-fr) [reply src/osd/scrubber_common.h:21]: Not sure if I've answered that one:
the ScrubberPasskey is an existing mechanism, used by the rest of the scrub-related objects.
As that number of those objects increased, the PG became too "friendly".

Note that all I did here is to move the Passkey to another '.h' file

**Comment by ronen-fr** [comment]
waiting for https://github.com/ceph/ceph/pull/46079 to be merged. I'll "un-draft" this one here - then.

**Comment by athanatos** [comment]
Failing make check

../src/crimson/osd/pg.h:464:26: error: no member named 'get_pool' in 'PeeringState'; did you mean 'get_pgpool'?
    return peering_state.get_pool();

**Comment by ronen-fr** [comment]
> Failing make check
> 
> ../src/crimson/osd/pg.h:464:26: error: no member named 'get_pool' in 'PeeringState'; did you mean 'get_pgpool'? return peering_state.get_pool();

@athanatos - yes. Seems I botched the Crimson patch. Fixing it now.

File: src/osd/osd_types_fmt.h (hunk 9 above)
Comment (ronen-fr) [reply src/osd/osd_types_fmt.h]: moved to a separate PR

File: src/test/osd/test_scrubber_be.cc (hunk 11 above)
Comment (ronen-fr) [reply src/test/osd/test_scrubber_be.cc]: done

File: src/test/osd/test_scrubber_be.cc (hunk 12 above)
Comment (ronen-fr) [reply src/test/osd/test_scrubber_be.cc]: It was more of an inspiration source.. The specific data will be changed later on, when the tests are improved. Didn't see a point in extracting that part.

File: src/osd/scrubber/scrub_backend.h [hunk 14]
```diff
@@ -83,6 +86,21 @@ struct error_counters_t {
   int deep_errors{0};
 };
 
+// the PgScrubber services used by the backend
+struct ScrubBeListener {
+  virtual std::ostream& gen_prefix(std::ostream& out) const = 0;
+  virtual CephContext* get_pg_cct() const = 0;
+  virtual CephContext* get_osd_cct() const = 0;
+  virtual LoggerSinkSet* get_logger() const = 0;
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.h:94]: Returns a reference here.

File: src/osd/scrubber/scrub_backend.h [hunk 15]
```diff
@@ -387,7 +401,7 @@ class ScrubBackend {
 
   // shorthands:
   ConfigProxy& m_conf;
-  LogChannelRef clog;
+  LoggerSinkSet* clog;
```
Comment (athanatos) [inline src/osd/scrubber/scrub_backend.h:404]: And store as a reference here.

**Comment by athanatos** [comment]
Audit uses of override for cases where final would be correct.  I don't think there are actually any cases where there are implemented methods that get overridden by another inheritor.

File: src/osd/PeeringState.h [hunk 16]
```diff
@@ -2136,7 +2153,7 @@ class PeeringState : public MissingLoc::MappingInfo {
     return last_rollback_info_trimmed_to_applied;
   }
   /// Returns stable reference to internal pool structure
-  const PGPool &get_pool() const {
+  const PGPool &get_pgpool() const {
```
Comment (athanatos) [inline src/osd/PeeringState.h:2156]: Did this need to change?

File: src/osd/scrubber_common.h [hunk 17]
```diff
@@ -29,6 +47,25 @@ struct ScrubPreconds {
   bool only_deadlined{false};
 };
 
+/// PG services used by the scrubber backend
+struct PgScrubBeListener {
+  virtual ~PgScrubBeListener() = default;
+
+  virtual const PGPool& get_pgpool() const = 0;
```
Comment (athanatos) [inline src/osd/scrubber_common.h:54]: Why not name this get_pool()?  I don't think it would be ambiguous.

File: src/osd/scrubber_common.h (hunk 7 above)
Comment (athanatos) [reply src/osd/scrubber_common.h:21]: As a reader/reviewer, it's just another type I have to lookup every time I see commits changing this portion of the code base.  The two interfaces you've defined already sufficiently clarify what portions of PG the scrubber objects can access, and this really just seems like extra cognitive overhead.

File: src/osd/ReplicatedBackend.h [hunk 18]
```diff
@@ -431,7 +431,11 @@ class ReplicatedBackend : public PGBackend {
     ScrubMap &map,
     ScrubMapBuilder &pos,
     ScrubMap::object &o) override;
-  uint64_t be_get_ondisk_size(uint64_t logical_size) override { return logical_size; }
+
+  uint64_t be_get_ondisk_size(uint64_t logical_size) const override
```
Comment (athanatos) [inline src/osd/ReplicatedBackend.h:435]: .h file, methods have opening brace on the same line as the signature.

Comment (athanatos) [reply src/osd/ReplicatedBackend.h:435]: Probably should be final?

File: src/osd/scrubber/pg_scrubber.h [hunk 19]
```diff
@@ -573,16 +574,28 @@ class PgScrubber : public ScrubPgIF,
...
 
-  virtual void submit_digest_fixes(const digests_fixes_t& fixes)
+  void submit_digest_fixes(const digests_fixes_t& fixes) override
   {
     ceph_assert(0 && "expecting a PrimaryLogScrub object");
   }
 
+  /// \todo clearly identify when to use which of the following two:
+  CephContext* get_pg_cct() const override { return m_pg->cct; }
+  CephContext* get_osd_cct() const override { return m_pg->cct; }
```
Comment (athanatos) [inline src/osd/scrubber/pg_scrubber.h:589]: As far as I know, it is not possible for those to be different.  The CephContext* instance should be process-wide.

File: src/test/osd/test_scrubber_be.cc [hunk 20]
```diff
@@ -0,0 +1,647 @@
...
+    ScrubGenerator::make_real_objs_conf(pool_id, real_objs, acting_osds);
+
+  // now we can create the main mockers
+
+  // the "PgScrubber"
+  test_scrubber = std::make_unique<TestScrubber>(spg, osdmap, &logger);
+
+  // the "PG" (and its backend)
+  test_pg = std::make_unique<TestPg>(pool, info, i_am);
+  std::cout << fmt::format("{}: test_pg: OK\n", __func__);
```
Comment (athanatos) [inline src/test/osd/test_scrubber_be.cc:329]: Not sure what this log line means.

File: src/test/osd/test_scrubber_be.cc [hunk 21]
```diff
@@ -0,0 +1,647 @@
...
+  std::cout << fmt::format("{}: acting: {}\n", __func__, acting_shards);
```
Comment (athanatos) [inline src/test/osd/test_scrubber_be.cc:330]: std::endl;  std::cout doesn't necessarily flush without a std::endl, and readers of this debugging will want to be able to rely on log output ordering.

**Comment by athanatos** [comment]
Interface/formatting quibbles aside, the testing infrastructure looks excellent -- seems like we should be able to represent a really broad set of possible mismatches!  Once the above are addresses, it should be good to go.

File: src/test/osd/test_scrubber_be.cc [hunk 22]
```diff
@@ -0,0 +1,647 @@
...
+  // info.last_osdmap_epoch = osdmap->get_epoch();
+  // info.history.last_epoch_marked_removed = osdmap->get_epoch();
+  info.last_user_version = 1;
+  info.purged_snaps = {};
+  info.last_user_version = 1;
+  info.history.last_epoch_clean = osdmap->get_epoch();
+  info.history.last_epoch_split = osdmap->get_epoch();
+  info.history.last_epoch_marked_full = osdmap->get_epoch();
+  info.last_backfill = hobject_t::get_max();
+  info.stats.stats.sum.num_objects_degraded = 0;
```
Comment (Matan-B) [inline src/test/osd/test_scrubber_be.cc:439]: `object_stat_sum_t()` constructor sets zero to it's members.

File: src/test/osd/test_scrubber_be.cc [hunk 23]
```diff
@@ -0,0 +1,647 @@
...
+  info.stats.stats.sum.num_objects = 0;
+  info.stats.stats.sum.num_object_clones = 0;
+  info.stats.stats.sum.num_object_copies = 0;
+  info.stats.stats.sum.num_objects_missing_on_primary = 0;
+  info.stats.stats.sum.num_objects_degraded = 0;
+  info.stats.stats.sum.num_objects_misplaced = 0;
+  info.stats.stats.sum.num_objects_unfound = 0;
+  // info.stats.stats.sum.num_bytes_used = 0; // FIXME: find the right value
+  info.stats.stats.sum.num_bytes = 0;
+  info.stats.stats.sum.num_objects = 0;
```
Comment (Matan-B) [inline src/test/osd/test_scrubber_be.cc:451]: `object_stat_sum_t()` constructor sets zero to it's members.

File: src/test/osd/scrubber_generators.cc [hunk 24]
```diff
@@ -0,0 +1,164 @@
...
+    return {bufferptr(), {}};
+  }
+  /// \todo fill in missing version/osd details from the robj
+  auto sns = snapset_mock_data->make_snapset();
+  bufferlist bl;
+  encode(sns, bl);
+  bufferptr bp = bufferptr(bl.c_str(), bl.length());
+
+  // extract the set of object snaps
+  return {bp, sns.snaps};
```
Comment (Matan-B) [inline src/test/osd/scrubber_generators.cc:37]: Nit: Can't we return `{bl, sns.snaps};`? 
line 73: `ret.smobj.attrs[SS_ATTR] = std::move(bl);`

Comment (ronen-fr) [reply src/test/osd/scrubber_generators.cc:37]: The 'bl' is a buffer-list, not a buffer pointer. What am I missing?

File: src/test/osd/test_scrubber_be.cc (hunk 23 above)
Comment (ronen-fr) [reply src/test/osd/test_scrubber_be.cc:451]: Thanks. (And later on I will modify osd_types.h to unify the initializations - currently some in the member dec and some in the ctor)

File: src/test/osd/test_scrubber_be.cc (hunk 21 above)
Comment (ronen-fr) [reply src/test/osd/test_scrubber_be.cc:330]: Will change (I was opting for flushing on selected lines, just to declutter the code).

File: src/osd/scrubber_common.h (hunk 17 above)
Comment (ronen-fr) [reply src/osd/scrubber_common.h:54]: For a time - it was, as I had structures that provided both APIS, i.e. both types of get_pool().
Seems that it's no longer so, but still I think the change makes sense: otherwise -
it is hardly evident from a piece of code what interface is used.
- and the number of actual callers that had to be changed is small.

File: src/osd/PeeringState.h (hunk 16 above)
Comment (ronen-fr) [reply src/osd/PeeringState.h:2156]: see next comment

File: src/osd/ReplicatedBackend.h (hunk 18 above)
Comment (ronen-fr) [reply src/osd/ReplicatedBackend.h:435]: Re the '.h' - I am using clang-format, and don't seem to have an easy way to
diff between 'cc' and 'h' files. And - I didn't see a clear reason for the distinction.
Anyway - will revert existing files manually.

File: src/osd/scrubber/scrub_backend.h (hunk 15 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.h:404]: That's a problem:
For example: take

LoggerSinkSet* PgScrubber::get_logger() const { return m_osds->clog.get(); }

I can't return a non-owning ref here (LoggerSinkSet&). And will have similar issues if
returning a owning ref to base.

File: src/osd/scrubber/scrub_backend.h (hunk 14 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.h:94]: See next comment

File: src/osd/scrubber/pg_scrubber.h (hunk 19 above)
Comment (ronen-fr) [reply src/osd/scrubber/pg_scrubber.h:589]: Thanks. Will merge.

File: src/osd/scrubber_common.h (hunk 7 above)
Comment (ronen-fr) [reply src/osd/scrubber_common.h:21]: Let's discuss this online. The passkey goal was "PG-oriented" - limiting the amount of code that
can potentially modify PG class internals (a different type of cognitive load). Not really related to
scrubber's code design.
Now that it's there - I don't think this PR is the right place to undo it.

Comment (athanatos) [reply src/osd/scrubber_common.h:21]: Ok

File: src/osd/scrubber_common.h (hunk 17 above)
Comment (athanatos) [reply src/osd/scrubber_common.h:54]: Ok, we can change it to get_pgpool() if there's a get_pool somewhere that returns a PGPool&.

File: src/osd/scrubber/scrub_backend.h (hunk 15 above)
Comment (athanatos) [reply src/osd/scrubber/scrub_backend.h:403]: https://github.com/athanatos/ceph/commit/30736134ce9dc56126b472a4338e6b0332e00e1f I think we can return it as a reference.  It's not *that* important, but I'd prefer to use references in non-owning structs if possible.

File: src/osd/scrubber_common.h (hunk 17 above)
Comment (ronen-fr) [reply src/osd/scrubber_common.h:54]: in PG.h it's:
`const PGPool& get_pool() const`

but in PrimaryLogPG (that inherits from it - but also from PGBackend::listener) it's
`const pg_pool_t &get_pool() const override`

Comment (athanatos) [reply src/osd/scrubber_common.h:54]: Oh!  Yeah, I'm with you, PG::get_pgpool() it is.

File: src/osd/scrubber/scrub_backend.h (hunk 15 above)
Comment (ronen-fr) [reply src/osd/scrubber/scrub_backend.h]: done

**Review by athanatos — APPROVED** [review]
LGTM pending test run

**Comment by ronen-fr** [comment]
Will be merged based on Teuthology results:
http://pulpito.front.sepia.ceph.com/?branch=wip-rf-sbe-test2
(and - to some degree - on
http://pulpito.front.sepia.ceph.com/?branch=wip-rf-sched-test
as wip-rf-sched-test is based on wip-rf-sbe-test2 with
additional unrelated two commits).

Failures in the tests:

- 4+4 failues in the 'standalone' suite: some 'centos' runs that fail to start;
  All OK when specifying Ubuntu.
- One dead job in rados:thrash (6834893): startup (ansible) issue;
- Rados suite - failed due to Teu issue. Rerun fine as wip-rf-sched-test.
- One failed rados:thrash (6832161) - OSD startup crash (in read_log_and_missing())
